src/App.js

```diff
@@ -1,15 +1,90 @@
 import SearchBar from './components/SearchBar/SearchBar';
 import Users from './components/Users/Users';
 import './App.css';
+import axios from 'axios';
+import { useEffect, useState } from 'react';
 
 function App() {
+  // hooks
+  const [users, setUsers] = useState([]);
+  const [searchInput, setSearchInput] = useState('');
+  const [expanded, setExpanded] = useState([]);
+  const [loading, setLoading] = useState(true);
+  const [error, setError] = useState('');
+
   // TODO: Fetch data here
+  useEffect(() => {
+    axios.get('https://users-app-backend.onrender.com/users')
+      .then((response) => {
+        setUsers(response.data.data);
+        if (response){
+          setLoading(false);
+        }
+      })
+      .catch((error) => {
+        setLoading(true);
+        setError("Error fetching data from API");
+      })
+  }, []);
+
+  // filter the users that match the `searchInput`
+  // the users data is an object
+  let filteredData = users;
+
+  if (searchInput){
+    filteredData = users.filter(user => {
+    const validInput = `${user.name} ${user.country} ${user.company}`;
+    const validInputToLowerCase = validInput.toLowerCase();
+    const searchInputToLowerCase = searchInput.toLowerCase();
+    return validInputToLowerCase.includes(searchInputToLowerCase);
+    });
+  }
+
+  // Expand All and Collapse All
+  const handleExpandAll = () => {
+    const allUserIds = users.map((user) => user.id);
+    setExpanded(allUserIds);
+  }
+
+  const handleCollapseAll = () => {
+    setExpanded([]);
+  }
+
+  // if loading state or error, don't display users. Else, display users.
+  const renderContent = () => {
+    if (loading){
+      return <div>Loading...</div>
+    } else if (error) {
+      return <div>{error}</div>
+    } else {
+      return <Users 
+          filteredData={filteredData}
+          expanded={expanded}
+          setExpanded={setExpanded}
+        />
+    }
+  }
+
+  // if user input doesn't create filtered data, show no results message
+  const renderNoResults = () => {
+    if (searchInput.length > 0 && filteredData.length === 0) {
+      return <div>No results for {searchInput}</div>
+    }
+  }
 
   return (
     <div className="App">
       <h1>Our Users</h1>
-      <SearchBar />
-      <Users />
+      <div className="inputExpandContainer">
+        <SearchBar 
+          searchInput={searchInput} 
+          setSearchInput={setSearchInput}
+        />
+        <button onClick={handleExpandAll} >Expand All</button>
+        <button onClick={handleCollapseAll} >Collapse All</button>
+      </div>
+      {renderNoResults()}
+      {renderContent()}
```

Comment:
I've seen this approach before where there are two functions (`renderNoResults()` and `renderContent()`). For this to work, you have to be sure that in every case, _only one_ function will return JSX. There can't be cases where _both_ functions return JSX. There also can't be any cases where _no_ functions return JSX. 

For this reason, this type of approach has always scared me a little bit. Right now it works, but another developer could misunderstand the intention and accidentally break this. There is already something semi-risky going on with `renderNoResults()` where we have to confirm that not only there is no data (which could also be true during the loading state) but there is also search input (which implies we are not loading). But let's say later on we have to implement a `Refresh` button that clears the current data and re-fetches the data from the API. If the user clicks this while there is input in the text box, the logic will break down. That's one example of what could go wrong.

My personal view is that if we want the code on lines 86-87 to always result in one and only one outcome, then we should combine them into one function that always returns exactly one thing. The principle behind this is that there should always be only one possible state of the application: either loading, error, results, or no results. It's a good idea to make sure there can't be combinations of states like loading and no results.

To implement this, you might be able to move the logic in `renderNoResults()` into `renderContent()`, where it would be another case. In this way `renderContent()` would hold all the logic. For example:

```
if loading => return <div>Loading...</div>
else if error => return <div>{error}</div>
else if filteredData.length > 0 => return <Users ... />
else return <div>No results for {searchInput}</div>
```

Reply:
Thank you! I just tried combining the renderNoResults() into the renderContent() and now there is only one function returning the JSX, and it still works for showing "no results for {input}"